File: src/key.rs
use bls12_381::{G2Affine, G2Projective, Gt, Scalar};

use group::{Curve, Group, ff::Field};
use rand::RngCore;
use subtle::CtOption;


const COMPRESSED_GT_SIZE: usize = 576 / 2;
const COMPRESSED_G2_SIZE: usize = 96;
const SCALAR_SIZE: usize = 32;

pub struct EncryptionPublicKey {
    pub(crate) za1: Gt,
}

impl EncryptionPublicKey {
    const BYTES: usize = COMPRESSED_GT_SIZE;

    pub fn to_bytes(&self) -> [u8; Self::BYTES] {
        return self.za1.serialize_compressed();
    }

    pub fn from_bytes(bytes: &[u8; Self::BYTES]) -> CtOption<Self> {
        Gt::deserialize_compressed(&bytes).map(|za1| Self { za1 })
    }
}

pub struct DecryptionPublicKey {
    pub(crate) gb2: G2Affine,
}

impl DecryptionPublicKey {
    const BYTES: usize = COMPRESSED_G2_SIZE;

    pub fn to_bytes(&self) -> [u8; Self::BYTES] {
        return self.gb2.to_compressed();
    }

    pub fn from_bytes(bytes: &[u8; Self::BYTES]) -> CtOption<Self> {
        G2Affine::from_compressed(bytes).map(|ga2| Self { gb2: ga2 })
    }
}

pub struct EncryptionSecretKey {
    pub(crate) a1: Scalar,
}

impl EncryptionSecretKey {
    pub fn pubkey(&self) -> EncryptionPublicKey {
        EncryptionPublicKey {
            za1: Gt::generator() * self.a1,
        }
    }

    pub fn to_bytes(&self) -> [u8; SCALAR_SIZE] {
        self.a1.to_bytes()
    }

    pub fn from_bytes(bytes: &[u8; SCALAR_SIZE]) -> CtOption<Self> {
        Scalar::from_bytes(bytes).map(|a1| Self { a1 })
    }

    pub fn random(rng: impl RngCore) -> Self {
        Self { a1: Scalar::random(rng) }
    }
}

pub struct DecryptionSecretKey {
    pub(crate) b2: Scalar,
}

impl DecryptionSecretKey {
    pub fn pubkey(&self) -> DecryptionPublicKey {
        DecryptionPublicKey {
            gb2: (G2Projective::generator() * self.b2).to_affine(),
        }
    }

    pub fn to_bytes(&self) -> [u8; SCALAR_SIZE] {
        self.b2.to_bytes()
    }

    pub fn from_bytes(bytes: &[u8; SCALAR_SIZE]) -> CtOption<Self> {
        Scalar::from_bytes(bytes).map(|b2| Self { b2 })
    }

    pub fn random(rng: impl RngCore) -> Self {
        Self { b2: Scalar::random(rng) }
    }
}

File: qafgh/key.go
package qafgh

import (
	"io"
	"math/big"

	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
	"github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
)

const scalarSize = 32

const EncryptionSKSize = scalarSize
const DecryptionSKSize = scalarSize

const EncryptionPKSize = compressedGtSize
const DecryptionPKSize = bls12381.SizeOfG2AffineCompressed

type EncryptionSecretKey struct {
	a1 fr.Element
}

func (sk *EncryptionSecretKey) ToBytes() [EncryptionSKSize]byte {
	return serScalar(&sk.a1)
}

func EncryptionSecretKeyFromBytes(inp []byte) (*EncryptionSecretKey, error) {
	s, err := dsrScalar(inp)
	if err != nil {
		return nil, err
	}
	return &EncryptionSecretKey{
		a1: s,
	}, nil
}

func (sk *EncryptionSecretKey) Pubkey() EncryptionPublicKey {
	res := GTZ // copy generator
	exp := big.Int{}
	sk.a1.ToBigIntRegular(&exp)
	res.Exp(&res, exp)
	return EncryptionPublicKey{
		za1: res,
	}
}

func RandomEncryptionSecretKey(rand io.Reader) (*EncryptionSecretKey, error) {
	a1, err := randScalar(rand)
	if err != nil {
		return nil, err
	}
	return &EncryptionSecretKey{a1: a1}, nil
}

type EncryptionPublicKey struct {
	za1 bls12381.GT
}

func (pk *EncryptionPublicKey) ToBytes() [EncryptionPKSize]byte {
	return compressGt(pk.za1)
}

func EncryptionPublicKeyFromBytes(inp []byte) (*EncryptionPublicKey, error) {
	za1, err := decompressGt(inp)
	if err != nil {
		return nil, err
	}
	return &EncryptionPublicKey{
		za1: *za1,
	}, nil
}

type DecryptionSecretKey struct {
	b2 fr.Element
}

func (sk *DecryptionSecretKey) ToBytes() [EncryptionSKSize]byte {
	return serScalar(&sk.b2)
}

func DecryptionSecretKeyFromBytes(inp []byte) (*DecryptionSecretKey, error) {
	s, err := dsrScalar(inp)
	if err != nil {
		return nil, err
	}
	return &DecryptionSecretKey{
		b2: s,
	}, nil
}

func (sk *DecryptionSecretKey) Pubkey() DecryptionPublicKey {
	res := bls12381.G2Affine{}
	b2Big := scalarToBig(&sk.b2)
	res.ScalarMultiplication(&G2GEN, &b2Big)
	return DecryptionPublicKey{
		gb2: res,
	}
}

func RandomDecryptionSecretKey(rand io.Reader) (*DecryptionSecretKey, error) {
	b2, err := randScalar(rand)
	if err != nil {
		return nil, err
	}
	return &DecryptionSecretKey{b2: b2}, nil
}

type DecryptionPublicKey struct {
	gb2 bls12381.G2Affine
}

func (pk *DecryptionPublicKey) ToBytes() [DecryptionPKSize]byte {
	return pk.gb2.Bytes()
}

func DecryptionPublicKeyFromBytes(inp []byte) (*DecryptionPublicKey, error) {
	gb2, err := dsrG2(inp)
	if err != nil {
		return nil, err
	}
	return &DecryptionPublicKey{
		gb2: gb2,
	}, nil
}
